rama-dns/src/lib.rs

```diff
@@ -79,6 +79,88 @@ impl<R: DnsResolver<Error: Into<BoxError>>> DnsResolver for Option<R> {
     }
 }
 
+#[derive(Debug)]
+pub struct DnsChainDomainResolveErr<E: 'static> {
+    errors: Vec<E>,
+}
+impl<E: std::fmt::Debug + Send> std::fmt::Display for DnsChainDomainResolveErr<E> {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "{:?}", self.errors)
+    }
+}
+
+impl<E: std::fmt::Debug + Send + std::error::Error> std::error::Error
+    for DnsChainDomainResolveErr<E>
+{
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        if self.errors.is_empty() {
```

Comment:
Pretty certain you can replace these lines with just a single call:

```
self.errors.last()
```

https://doc.rust-lang.org/std/vec/struct.Vec.html#method.last

(we want to return the last error btw, as source, given that's the reason why it failed in the end, as up to then it was still possible to have success)